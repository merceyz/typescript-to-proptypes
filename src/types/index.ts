export * from './nodes/baseNodes';
export * from './nodes/program';
export * from './nodes/component';
export * from './nodes/proptype';

export * from './props/function';
export * from './props/interface';
export * from './props/string';
export * from './props/union';
export * from './props/undefined';
export * from './props/boolean';
export * from './props/numeric';
export * from './props/literal';
export * from './props/any';
export * from './props/object';
export * from './props/array';
export * from './props/element';
export * from './props/instanceOf';
export * from './props/DOMElement';
