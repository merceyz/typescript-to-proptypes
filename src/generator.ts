import * as t from './types';
import _ from 'lodash';

export interface GenerateOptions {
	/**
	 * Enable/disable the default sorting (ascending) or provide your own sort function
	 * @default true
	 */
	sortProptypes?: boolean | ((a: t.PropTypeNode, b: t.PropTypeNode) => 0 | -1 | 1);

	/**
	 * The name used when importing prop-types
	 * @default 'PropTypes'
	 */
	importedName?: string;

	/**
	 * Enable/disable including JSDoc comments
	 * @default true
	 */
	includeJSDoc?: boolean;

	/**
	 * Control which PropTypes are included in the final result
	 * @param proptype The current PropType about to be converted to text
	 */
	shouldInclude?(proptype: t.PropTypeNode): boolean | undefined;

	/**
	 * A comment that will be added to the start of the PropTypes code block
	 * @example
	 * foo.propTypes = {
	 *  // Comment goes here
	 * }
	 */
	comment?: string;
}

/**
 * Generates code from the given node
 * @param node The node to convert to code
 * @param options The options used to control the way the code gets generated
 */
export function generate(node: t.Node | t.PropTypeNode[], options: GenerateOptions = {}): string {
	const {
		sortProptypes = true,
		importedName = 'PropTypes',
		includeJSDoc = true,
		shouldInclude,
	} = options;

	function jsDoc(node: t.PropTypeNode | t.LiteralNode) {
		if (!includeJSDoc || !node.jsDoc) {
			return '';
		}
		return `/**\n* ${node.jsDoc.split(/\r?\n/).reduce((prev, curr) => prev + '\n* ' + curr)}\n*/\n`;
	}

	if (Array.isArray(node)) {
		let propTypes = node;

		if (typeof sortProptypes === 'function') {
			propTypes = propTypes.sort(sortProptypes);
		} else if (sortProptypes === true) {
			propTypes = propTypes.sort((a, b) => a.name.localeCompare(b.name));
		}

		let filteredNodes = node;
		if (shouldInclude) {
			filteredNodes = filteredNodes.filter((x) => shouldInclude(x));
		}

		return filteredNodes
			.map((prop) => generate(prop, options))
			.reduce((prev, curr) => `${prev}\n${curr}`);
	}

	if (t.isProgramNode(node)) {
		return node.body
			.map((prop) => generate(prop, options))
			.reduce((prev, curr) => `${prev}\n${curr}`);
	}

	if (t.isComponentNode(node)) {
		const comment =
			options.comment &&
			`// ${options.comment.split(/\r?\n/gm).reduce((prev, curr) => `${prev}\n// ${curr}`)}\n`;

		return `${node.name}.propTypes = {\n${comment ? comment : ''}${generate(
			node.types,
			options
		)}\n}`;
	}

	if (t.isPropTypeNode(node)) {
		let isOptional = false;
		let propType = { ...node.propType };

		if (t.isUnionNode(propType) && propType.types.some(t.isUndefinedNode)) {
			isOptional = true;
			propType.types = propType.types.filter(
				(prop) => !t.isUndefinedNode(prop) && !(t.isLiteralNode(prop) && prop.value === 'null')
			);
			if (propType.types.length === 1 && t.isLiteralNode(propType.types[0]) === false) {
				propType = propType.types[0];
			}
		}

		return `${jsDoc(node)}"${node.name}": ${generate(propType, options)}${
			isOptional ? '' : '.isRequired'
		},`;
	}

	if (t.isInterfaceNode(node)) {
		return `${importedName}.shape({\n${generate(node.types, {
			...options,
			shouldInclude: undefined,
		})}\n})`;
	}

	if (t.isFunctionNode(node)) {
		return `${importedName}.func`;
	}

	if (t.isStringNode(node)) {
		return `${importedName}.string`;
	}

	if (t.isBooleanNode(node)) {
		return `${importedName}.bool`;
	}

	if (t.isNumericNode(node)) {
		return `${importedName}.number`;
	}

	if (t.isLiteralNode(node)) {
		return `${importedName}.oneOf([${jsDoc(node)}${node.value}])`;
	}

	if (t.isObjectNode(node)) {
		return `${importedName}.object`;
	}

	if (t.isAnyNode(node)) {
		return `${importedName}.any`;
	}

	if (t.isElementNode(node)) {
		return `${importedName}.${node.elementType}`;
	}

	if (t.isInstanceOfNode(node)) {
		return `${importedName}.instanceOf(${node.instance})`;
	}

	if (t.isArrayNode(node)) {
		if (t.isAnyNode(node.arrayType)) {
			return `${importedName}.array`;
		}

		return `${importedName}.arrayOf(${generate(node.arrayType, options)})`;
	}

	if (t.isUnionNode(node)) {
		let [literals, rest] = _.partition(node.types, t.isLiteralNode);
		literals = _.uniqBy(literals, (x) => x.value);
		rest = _.uniqBy(rest, (x) => (t.isInstanceOfNode(x) ? `${x.type}.${x.instance}` : x.type));

		literals = literals.sort((a, b) => a.value.localeCompare(b.value));

		const nodeToStringName = (obj: t.Node): string => {
			if (t.isInstanceOfNode(obj)) {
				return `${obj.type}.${obj.instance}`;
			} else if (t.isInterfaceNode(obj)) {
				// An interface is PropTypes.shape
				// Use `ShapeNode` to get it sorted in the correct order
				return `ShapeNode`;
			}

			return obj.type;
		};

		rest = rest.sort((a, b) => nodeToStringName(a).localeCompare(nodeToStringName(b)));

		if (literals.find((x) => x.value === 'true') && literals.find((x) => x.value === 'false')) {
			rest.push(t.booleanNode());
			literals = literals.filter((x) => x.value !== 'true' && x.value !== 'false');
		}

		const literalProps =
			literals.length !== 0
				? `${importedName}.oneOf([${literals
						.map((x) => `${jsDoc(x)}${x.value}`)
						.reduce((prev, curr) => `${prev},${curr}`)}])`
				: '';

		if (rest.length === 0) {
			return literalProps;
		}

		if (literals.length === 0 && rest.length === 1) {
			return generate(rest[0], options);
		}

		return `${importedName}.oneOfType([${literalProps ? literalProps + ', ' : ''}${rest
			.map((x) => generate(x, options))
			.reduce((prev, curr) => `${prev},${curr}`)}])`;
	}

	throw new Error(`Nothing to handle node of type "${node.type}"`);
}
