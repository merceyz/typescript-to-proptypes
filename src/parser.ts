import * as ts from 'typescript';
import * as t from './types';
import * as doctrine from 'doctrine';

/**
 * Options that specify how the parser should act
 */
export interface ParserOptions {
	/**
	 * Called before a PropType is added to a component/object
	 * @return true to include the PropType, false to skip it, or undefined to
	 * use the default behaviour
	 * @default name !== 'ref'
	 */
	shouldInclude: (data: { name: string; depth: number }) => boolean | undefined;
	/**
	 * Called before the shape of an object is resolved
	 * @return true to resolve the shape of the object, false to just use a object, or undefined to
	 * use the default behaviour
	 * @default propertyCount <= 50 && depth <= 3
	 */
	shouldResolveObject: (data: {
		name: string;
		propertyCount: number;
		depth: number;
	}) => boolean | undefined;
	/**
	 * Control if const declarations should be checked
	 * @default false
	 * @example declare const Component: React.ComponentType<Props>;
	 */
	checkDeclarations?: boolean;
}

/**
 * A wrapper for `ts.createProgram`
 * @param files The files to later be parsed with `parseFromProgram`
 * @param options The options to pass to the compiler
 */
export function createProgram(files: string[], options: ts.CompilerOptions) {
	return ts.createProgram(files, options);
}

/**
 * Creates a program, parses the specified file and returns the PropTypes as an AST, if you need to parse more than one file
 * use `createProgram` and `parseFromProgram` for better performance
 * @param filePath The file to parse
 * @param options The options from `loadConfig`
 * @param parserOptions Options that specify how the parser should act
 */
export function parseFile(
	filePath: string,
	options: ts.CompilerOptions,
	parserOptions: Partial<ParserOptions> = {}
) {
	const program = ts.createProgram([filePath], options);
	return parseFromProgram(filePath, program, parserOptions);
}

/**
 * Parses the specified file and returns the PropTypes as an AST
 * @param filePath The file to get the PropTypes from
 * @param program The program object returned by `createProgram`
 * @param parserOptions Options that specify how the parser should act
 */
export function parseFromProgram(
	filePath: string,
	program: ts.Program,
	parserOptions: Partial<ParserOptions> = {}
) {
	const { checkDeclarations = false } = parserOptions;

	const shouldInclude: ParserOptions['shouldInclude'] = (data) => {
		if (parserOptions.shouldInclude) {
			const result = parserOptions.shouldInclude(data);
			if (result !== undefined) {
				return result;
			}
		}

		return data.name !== 'ref';
	};

	const shouldResolveObject: ParserOptions['shouldResolveObject'] = (data) => {
		if (parserOptions.shouldResolveObject) {
			const result = parserOptions.shouldResolveObject(data);
			if (result !== undefined) {
				return result;
			}
		}

		return data.propertyCount <= 50 && data.depth <= 3;
	};

	const checker = program.getTypeChecker();
	const sourceFile = program.getSourceFile(filePath);

	const programNode = t.programNode();
	const reactImports: string[] = [];

	if (sourceFile) {
		ts.forEachChild(sourceFile, visitImports);
		ts.forEachChild(sourceFile, visit);
	} else {
		throw new Error(`Program doesn't contain file "${filePath}"`);
	}

	return programNode;

	function visitImports(node: ts.Node) {
		if (
			ts.isImportDeclaration(node) &&
			ts.isStringLiteral(node.moduleSpecifier) &&
			node.moduleSpecifier.text === 'react' &&
			node.importClause
		) {
			const imports = ['Component', 'PureComponent', 'memo', 'forwardRef'];

			// import x from 'react'
			if (node.importClause.name) {
				const nameText = node.importClause.name.text;
				reactImports.push(...imports.map((x) => `${nameText}.${x}`));
			}

			// import {x, y as z} from 'react'
			const bindings = node.importClause.namedBindings;
			if (bindings) {
				if (ts.isNamedImports(bindings)) {
					bindings.elements.forEach((spec) => {
						const nameIdentifier = spec.propertyName || spec.name;
						const nameText = nameIdentifier.getText();
						if (imports.includes(nameText)) {
							reactImports.push(spec.name.getText());
						}
					});
				}
				// import * as x from 'react'
				else {
					const nameText = bindings.name.text;
					reactImports.push(...imports.map((x) => `${nameText}.${x}`));
				}
			}
		}
	}

	function visit(node: ts.Node) {
		// function x(props: type) { return <div/> }
		if (ts.isFunctionDeclaration(node) && node.name && node.parameters.length === 1) {
			parseFunctionComponent(node);
		}
		// const x = ...
		else if (ts.isVariableStatement(node)) {
			ts.forEachChild(node.declarationList, (variableNode) => {
				// x = (props: type) => { return <div/> }
				// x = function(props: type) { return <div/> }
				// x = function y(props: type) { return <div/> }
				// x = react.memo((props:type) { return <div/> })

				if (ts.isVariableDeclaration(variableNode) && variableNode.name) {
					const type = checker.getTypeAtLocation(variableNode.name);
					if (!variableNode.initializer) {
						if (
							checkDeclarations &&
							type.aliasSymbol &&
							type.aliasTypeArguments &&
							checker.getFullyQualifiedName(type.aliasSymbol) === 'React.ComponentType'
						) {
							parsePropsType(variableNode.name.getText(), type.aliasTypeArguments[0]);
						}
					} else if (
						(ts.isArrowFunction(variableNode.initializer) ||
							ts.isFunctionExpression(variableNode.initializer)) &&
						variableNode.initializer.parameters.length === 1
					) {
						parseFunctionComponent(variableNode);
					}
					// x = react.memo((props:type) { return <div/> })
					else if (
						ts.isCallExpression(variableNode.initializer) &&
						variableNode.initializer.arguments.length > 0
					) {
						const callString = variableNode.initializer.expression.getText();
						const arg = variableNode.initializer.arguments[0];
						if (
							reactImports.includes(callString) &&
							(ts.isArrowFunction(arg) || ts.isFunctionExpression(arg)) &&
							arg.parameters.length > 0
						) {
							const symbol = checker.getSymbolAtLocation(arg.parameters[0].name);
							if (symbol) {
								parsePropsType(
									variableNode.name.getText(),
									checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration)
								);
							}
						}
					}
				}
			});
		} else if (
			ts.isClassDeclaration(node) &&
			node.name &&
			node.heritageClauses &&
			node.heritageClauses.length === 1
		) {
			const heritage = node.heritageClauses[0];
			if (heritage.types.length !== 1) return;

			const arg = heritage.types[0];
			if (!arg.typeArguments) return;

			if (reactImports.includes(arg.expression.getText())) {
				parsePropsType(node.name.getText(), checker.getTypeAtLocation(arg.typeArguments[0]));
			}
		}
	}

	function isTypeJSXElementLike(type: ts.Type): boolean {
		if (type.isUnion()) {
			return type.types.every(
				(subType) => subType.flags & ts.TypeFlags.Null || isTypeJSXElementLike(subType)
			);
		} else if (type.symbol) {
			const name = checker.getFullyQualifiedName(type.symbol);
			return name === 'global.JSX.Element' || name === 'React.ReactElement';
		}

		return false;
	}

	function parseFunctionComponent(node: ts.VariableDeclaration | ts.FunctionDeclaration) {
		if (!node.name) {
			return;
		}

		const symbol = checker.getSymbolAtLocation(node.name);
		if (!symbol) {
			return;
		}

		const signature = checker
			.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration)
			.getCallSignatures()[0];

		if (!isTypeJSXElementLike(signature.getReturnType())) {
			return;
		}

		const type = checker.getTypeOfSymbolAtLocation(
			signature.parameters[0],
			signature.parameters[0].valueDeclaration
		);

		parsePropsType(node.name.getText(), type);
	}

	function parsePropsType(name: string, type: ts.Type) {
		const properties = type
			.getProperties()
			.filter((symbol) => shouldInclude({ name: symbol.getName(), depth: 1 }));
		if (properties.length === 0) {
			return;
		}

		programNode.body.push(
			t.componentNode(
				name,
				properties.map((x) => checkSymbol(x, [(type as any).id]))
			)
		);
	}

	function checkSymbol(symbol: ts.Symbol, typeStack: number[]): t.PropTypeNode {
		const declarations = symbol.getDeclarations();
		const declaration = declarations && declarations[0];

		// TypeChecker keeps the name for
		// { a: React.ElementType, b: React.ReactElement | boolean }
		// but not
		// { a?: React.ElementType, b: React.ReactElement }
		// get around this by not using the TypeChecker
		if (
			declaration &&
			ts.isPropertySignature(declaration) &&
			declaration.type &&
			ts.isTypeReferenceNode(declaration.type)
		) {
			const name = declaration.type.typeName.getText();
			if (
				name === 'React.ElementType' ||
				name === 'React.ComponentType' ||
				name === 'React.ReactElement'
			) {
				const elementNode = t.elementNode(
					name === 'React.ReactElement' ? 'element' : 'elementType'
				);

				return t.propTypeNode(
					symbol.getName(),
					getDocumentation(symbol),
					declaration.questionToken ? t.unionNode([t.undefinedNode(), elementNode]) : elementNode
				);
			}
		}

		const type = declaration
			? // The proptypes aren't detailed enough that we need all the different combinations
			  // so we just pick the first and ignore the rest
			  checker.getTypeOfSymbolAtLocation(symbol, declaration)
			: // The properties of Record<..., ...> don't have a declaration, but the symbol has a type property
			  ((symbol as any).type as ts.Type);

		if (!type) {
			throw new Error('No types found');
		}

		// Typechecker only gives the type "any" if it's present in a union
		// This means the type of "a" in {a?:any} isn't "any | undefined"
		// So instead we check for the questionmark to detect optional types
		let parsedType: t.Node | undefined = undefined;
		if (
			(type.flags & ts.TypeFlags.Any || type.flags & ts.TypeFlags.Unknown) &&
			declaration &&
			ts.isPropertySignature(declaration)
		) {
			parsedType = declaration.questionToken
				? t.unionNode([t.undefinedNode(), t.anyNode()])
				: t.anyNode();
		} else {
			parsedType = checkType(type, typeStack, symbol.getName());
		}

		return t.propTypeNode(symbol.getName(), getDocumentation(symbol), parsedType);
	}

	function checkType(type: ts.Type, typeStack: number[], name: string): t.Node {
		// If the typeStack contains type.id we're dealing with an object that references itself.
		// To prevent getting stuck in an infinite loop we just set it to an objectNode
		if (typeStack.includes((type as any).id)) {
			return t.objectNode();
		}

		{
			const typeNode = type as any;

			const symbol = typeNode.aliasSymbol ? typeNode.aliasSymbol : typeNode.symbol;
			const typeName = symbol ? checker.getFullyQualifiedName(symbol) : null;
			switch (typeName) {
				case 'global.JSX.Element':
				case 'React.ReactElement': {
					return t.elementNode('element');
				}
				case 'React.ElementType': {
					return t.elementNode('elementType');
				}
				case 'React.ReactNode': {
					return t.unionNode([t.elementNode('node'), t.undefinedNode()]);
				}
				case 'React.Component': {
					return t.instanceOfNode(typeName);
				}
				case 'Element':
				case 'HTMLElement': {
					return t.DOMElementNode();
				}
			}
		}

		// @ts-ignore - Private method
		if (checker.isArrayType(type)) {
			// @ts-ignore - Private method
			const arrayType: ts.Type = checker.getElementTypeOfArrayType(type);
			return t.arrayNode(checkType(arrayType, typeStack, name));
		}

		if (type.isUnion()) {
			const node = t.unionNode(type.types.map((x) => checkType(x, typeStack, name)));

			return node.types.length === 1 ? node.types[0] : node;
		}

		if (type.flags & ts.TypeFlags.String) {
			return t.stringNode();
		}

		if (type.flags & ts.TypeFlags.Number) {
			return t.numericNode();
		}

		if (type.flags & ts.TypeFlags.Undefined) {
			return t.undefinedNode();
		}

		if (type.flags & ts.TypeFlags.Any || type.flags & ts.TypeFlags.Unknown) {
			return t.anyNode();
		}

		if (type.flags & ts.TypeFlags.Literal) {
			if (type.isLiteral()) {
				return t.literalNode(
					type.isStringLiteral() ? `"${type.value}"` : type.value,
					getDocumentation(type.symbol)
				);
			}
			return t.literalNode(checker.typeToString(type));
		}

		if (type.flags & ts.TypeFlags.Null) {
			return t.literalNode('null');
		}

		if (type.getCallSignatures().length) {
			return t.functionNode();
		}

		// Object-like type
		{
			const properties = type.getProperties();
			if (properties.length) {
				if (
					shouldResolveObject({ name, propertyCount: properties.length, depth: typeStack.length })
				) {
					const filtered = properties.filter((symbol) =>
						shouldInclude({ name: symbol.getName(), depth: typeStack.length + 1 })
					);
					if (filtered.length > 0) {
						return t.interfaceNode(
							filtered.map((x) => checkSymbol(x, [...typeStack, (type as any).id]))
						);
					}
				}

				return t.objectNode();
			}
		}

		// Object without properties or object keyword
		if (
			type.flags & ts.TypeFlags.Object ||
			(type.flags & ts.TypeFlags.NonPrimitive && checker.typeToString(type) === 'object')
		) {
			return t.objectNode();
		}

		console.warn(
			`Unable to handle node of type "ts.TypeFlags.${ts.TypeFlags[type.flags]}", using any`
		);
		return t.anyNode();
	}

	function getDocumentation(symbol?: ts.Symbol): string | undefined {
		if (!symbol) {
			return undefined;
		}

		const decl = symbol.getDeclarations();
		if (decl) {
			// @ts-ignore - Private method
			const comments = ts.getJSDocCommentsAndTags(decl[0]) as any[];
			if (comments && comments.length === 1) {
				const commentNode = comments[0];
				if (ts.isJSDoc(commentNode)) {
					return doctrine.unwrapComment(commentNode.getText()).trim();
				}
			}
		}

		const comment = ts.displayPartsToString(symbol.getDocumentationComment(checker));
		return comment ? comment : undefined;
	}
}
