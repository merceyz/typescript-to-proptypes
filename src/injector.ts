import * as babel from '@babel/core';
import * as babelTypes from '@babel/types';
import * as t from './types/index';
import { generate, GenerateOptions } from './generator';
import { v4 as uuid } from 'uuid';

export type InjectOptions = {
	/**
	 * By default all unused props are omitted from the result.
	 * Set this to true to include them instead.
	 */
	includeUnusedProps?: boolean;
	/**
	 * By default existing PropTypes are left alone, set this to true
	 * to have them removed before injecting the PropTypes
	 */
	removeExistingPropTypes?: boolean;
	/**
	 * Used to control which props are includes in the result
	 * @return true to include the prop, false to skip it, or undefined to
	 * use the default behaviour
	 * @default includeUnusedProps ? true : data.usedProps.includes(`${data.prop.name}`)
	 */
	shouldInclude?(data: { prop: t.PropTypeNode; usedProps: string[] }): boolean | undefined;
} & Pick<GenerateOptions, 'sortProptypes' | 'includeJSDoc' | 'comment'>;

/**
 * Injects the PropTypes from `parse` into the provided JavaScript code
 * @param propTypes Result from `parse` to inject into the JavaScript code
 * @param target The JavaScript code to add the PropTypes to
 * @param options Options controlling the final result
 */
export function inject(
	propTypes: t.ProgramNode,
	target: string,
	options: InjectOptions = {}
): string | null {
	if (propTypes.body.length === 0) {
		return target;
	}

	const propTypesToInject = new Map<string, string>();

	const result = babel.transformSync(target, {
		plugins: [
			require.resolve('@babel/plugin-syntax-class-properties'),
			require.resolve('@babel/plugin-syntax-jsx'),
			plugin(propTypes, options, propTypesToInject),
		],
		configFile: false,
		babelrc: false,
		retainLines: true,
	});

	let code = result && result.code;
	if (!code) {
		return null;
	}

	// Replace the placeholders with the generated prop-types
	// Workaround for issues with comments getting removed and malformed
	propTypesToInject.forEach((value, key) => {
		code = code!.replace(key, `\n\n${value}\n\n`);
	});

	return code;
}

function plugin(
	propTypes: t.ProgramNode,
	options: InjectOptions = {},
	mapOfPropTypes: Map<string, string>
): babel.PluginObj {
	const { includeUnusedProps = false, removeExistingPropTypes = false, ...otherOptions } = options;

	const shouldInclude: InjectOptions['shouldInclude'] = (data) => {
		if (options.shouldInclude) {
			const result = options.shouldInclude(data);
			if (result !== undefined) {
				return result;
			}
		}

		return includeUnusedProps ? true : data.usedProps.includes(`${data.prop.name}`);
	};

	let importName = '';
	let needImport = false;
	let alreadyImported = false;

	return {
		visitor: {
			Program: {
				enter(path) {
					if (
						!path.node.body.some((n) => {
							if (
								babelTypes.isImportDeclaration(n) &&
								n.source.value === 'prop-types' &&
								n.specifiers.length
							) {
								importName = n.specifiers[0].local.name;
								alreadyImported = true;
								return true;
							}
						})
					) {
						importName = 'PropTypes';
					}

					if (removeExistingPropTypes) {
						path.get('body').forEach((nodePath) => {
							const { node } = nodePath;
							if (
								babelTypes.isExpressionStatement(node) &&
								babelTypes.isAssignmentExpression(node.expression, { operator: '=' }) &&
								babelTypes.isMemberExpression(node.expression.left) &&
								babelTypes.isIdentifier(node.expression.left.property, { name: 'propTypes' })
							) {
								nodePath.remove();
							}
						});
					}
				},
				exit(path) {
					if (alreadyImported || !needImport) return;

					const propTypesImport = babel.template.ast(
						`import ${importName} from 'prop-types'`
					) as babel.types.ImportDeclaration;

					const firstImport = path
						.get('body')
						.find((nodePath) => babelTypes.isImportDeclaration(nodePath.node));

					// Insert import after the first one to avoid issues with comment flags
					if (firstImport) {
						firstImport.insertAfter(propTypesImport);
					} else {
						path.node.body = [propTypesImport, ...path.node.body];
					}
				},
			},
			FunctionDeclaration(path) {
				const { node } = path;

				// Prevent visiting again
				if ((node as any).hasBeenVisited) {
					path.skip();
					return;
				}

				if (!node.id) return;
				const props = propTypes.body.find((prop) => prop.name === node.id!.name);
				if (!props) return;

				let usedProps: string[] = [];

				if (!includeUnusedProps) {
					const prop = node.params[0];
					if (babelTypes.isIdentifier(prop) || babelTypes.isObjectPattern(prop)) {
						usedProps = getUsedProps(path, prop);
					}
				}

				if (usedProps.length === 0 && !includeUnusedProps) return;

				needImport = true;

				// Prevent visiting again
				(node as any).hasBeenVisited = true;
				path.skip();

				injectPropTypes({
					nodeName: node.id.name,
					usedProps,
					path,
					props,
				});
			},
			VariableDeclarator(path) {
				const { node } = path;

				// Prevent visiting again
				if ((node as any).hasBeenVisited) {
					path.skip();
					return;
				}

				if (!babelTypes.isIdentifier(node.id)) return;
				const nodeName = node.id.name;

				const props = propTypes.body.find((prop) => prop.name === nodeName);
				if (!props) return;

				if (
					babelTypes.isArrowFunctionExpression(node.init) ||
					babelTypes.isFunctionExpression(node.init)
				) {
					getFromProp(node.init.params[0]);
				}
				// x = react.memo(props => <div/>)
				else if (babelTypes.isCallExpression(node.init)) {
					const arg = node.init.arguments[0];
					if (babelTypes.isArrowFunctionExpression(arg) || babelTypes.isFunctionExpression(arg)) {
						getFromProp(arg.params[0]);
					}
				}

				function getFromProp(prop: babelTypes.Node) {
					const usedProps =
						!includeUnusedProps &&
						(babelTypes.isIdentifier(prop) || babelTypes.isObjectPattern(prop))
							? getUsedProps(path, prop)
							: [];

					if (usedProps.length === 0 && !includeUnusedProps) return;

					needImport = true;

					// Prevent visiting again
					(node as any).hasBeenVisited = true;
					path.skip();

					injectPropTypes({ path: path.parentPath, usedProps, props: props!, nodeName });
				}
			},
			ClassDeclaration(path) {
				const { node } = path;

				// Prevent visiting again
				if ((node as any).hasBeenVisited) {
					path.skip();
					return;
				}

				if (!babelTypes.isIdentifier(node.id)) return;
				const nodeName = node.id.name;

				const props = propTypes.body.find((prop) => prop.name === nodeName);
				if (!props) return;

				const usedProps = !includeUnusedProps ? getUsedProps(path, undefined) : [];

				if (usedProps.length === 0 && !includeUnusedProps) return;

				needImport = true;

				// Prevent visiting again
				(node as any).hasBeenVisited = true;
				path.skip();

				injectPropTypes({
					nodeName,
					usedProps,
					path,
					props,
				});
			},
		},
	};

	function injectPropTypes(options: {
		path: babel.NodePath;
		usedProps: string[];
		props: t.ComponentNode;
		nodeName: string;
	}) {
		const { path, props, usedProps, nodeName } = options;

		const source = generate(props, {
			...otherOptions,
			importedName: importName,
			shouldInclude: (prop) => shouldInclude!({ prop, usedProps }),
		});

		const placeholder = `const a${uuid().replace(/\-/g, '_')} = null;`;

		mapOfPropTypes.set(placeholder, source);

		// Insert prop types
		if (babelTypes.isExportNamedDeclaration(path.parent)) {
			path.insertAfter(babel.template.ast(`export { ${nodeName} };`));
			path.insertAfter(babel.template.ast(placeholder));
			path.parentPath.replaceWith(path.node);
		} else if (babelTypes.isExportDefaultDeclaration(path.parent)) {
			path.insertAfter(babel.template.ast(`export default ${nodeName};`));
			path.insertAfter(babel.template.ast(placeholder));
			path.parentPath.replaceWith(path.node);
		} else {
			path.insertAfter(babel.template.ast(placeholder));
		}
	}
}

/**
 * Gets used props from path
 * @param rootPath The path to search for uses of rootNode
 * @param rootNode The node to start the search, if undefined searches for `this.props`
 */
function getUsedProps(
	rootPath: babel.NodePath,
	rootNode: babelTypes.ObjectPattern | babelTypes.Identifier | undefined
) {
	const usedProps: string[] = [];
	getUsedPropsInternal(rootNode);
	return usedProps;

	function getUsedPropsInternal(
		node: babelTypes.ObjectPattern | babelTypes.Identifier | undefined
	) {
		if (node && babelTypes.isObjectPattern(node)) {
			node.properties.forEach((x) => {
				if (babelTypes.isObjectProperty(x)) {
					if (babelTypes.isStringLiteral(x.key)) {
						usedProps.push(x.key.value);
					} else {
						usedProps.push(x.key.name);
					}
				} else if (babelTypes.isIdentifier(x.argument)) {
					getUsedPropsInternal(x.argument);
				}
			});
		} else {
			rootPath.traverse({
				VariableDeclarator(path) {
					const init = path.node.init;
					if (
						(node
							? babelTypes.isIdentifier(init, { name: node.name })
							: babelTypes.isMemberExpression(init) &&
							  babelTypes.isThisExpression(init.object) &&
							  babelTypes.isIdentifier(init.property, { name: 'props' })) &&
						babelTypes.isObjectPattern(path.node.id)
					) {
						getUsedPropsInternal(path.node.id);
					}
				},
				MemberExpression(path) {
					if (
						(node
							? babelTypes.isIdentifier(path.node.object, { name: node.name })
							: babelTypes.isMemberExpression(path.node.object) &&
							  babelTypes.isMemberExpression(path.node.object.object) &&
							  babelTypes.isThisExpression(path.node.object.object.object) &&
							  babelTypes.isIdentifier(path.node.object.object.property, { name: 'props' })) &&
						babelTypes.isIdentifier(path.node.property)
					) {
						usedProps.push(path.node.property.name);
					}
				},
			});
		}
	}
}
