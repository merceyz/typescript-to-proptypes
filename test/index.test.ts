import path from 'path';
import fs from 'fs';

import glob from 'glob';
import prettier from 'prettier';
import * as ttp from '../src';
import { TestOptions } from './types';

const prettierConfig = prettier.resolveConfig.sync(path.join(__dirname, '../.prettierrc'));

const testCases = glob.sync('**/input.{d.ts,ts,tsx}', { absolute: true, cwd: __dirname });

// Create program for all files to speed up tests
const program = ttp.createProgram(
	testCases,
	ttp.loadConfig(path.resolve(__dirname, '../tsconfig.json'))
);

for (const testCase of testCases) {
	const dirname = path.dirname(testCase);
	const testName = dirname.substr(__dirname.length + 1);
	const astPath = path.join(dirname, 'output.json');
	const outputPath = path.join(dirname, 'output.js');
	const optionsPath = path.join(dirname, 'options.ts');
	const inputJS = path.join(dirname, 'input.js');

	it(testName, () => {
		const options: TestOptions = fs.existsSync(optionsPath) ? require(optionsPath).default : {};

		const ast = ttp.parseFromProgram(testCase, program, options.parser);

		//#region Check AST matches
		// propsFilename will be different depending on where the project is on disk
		// Manually check that it's correct and then delete it
		const newAST = ttp.programNode(
			ast.body.map((component) => {
				expect(component.propsFilename).toBe(testCase);
				return { ...component, propsFilename: undefined };
			})
		);

		if (fs.existsSync(astPath)) {
			expect(newAST).toMatchObject(JSON.parse(fs.readFileSync(astPath, 'utf8')));
		} else {
			fs.writeFileSync(
				astPath,
				prettier.format(JSON.stringify(newAST), {
					...prettierConfig,
					filepath: astPath,
				})
			);
		}
		//#endregion

		let inputSource = null;
		if (testCase.endsWith('.d.ts')) {
			try {
				inputSource = fs.readFileSync(inputJS, { encoding: 'utf8' });
			} catch (error) {}
		} else {
			inputSource = ttp.ts.transpileModule(fs.readFileSync(testCase, 'utf8'), {
				compilerOptions: {
					target: ttp.ts.ScriptTarget.ESNext,
					jsx: ttp.ts.JsxEmit.Preserve,
				},
			}).outputText;
		}

		let result = '';
		// For d.ts files we just generate the AST
		if (!inputSource) {
			result = ttp.generate(ast, options.generator);
		}
		// For .tsx? files we transpile them and inject the proptypes
		else {
			const injected = ttp.inject(ast, inputSource, options.injector);
			if (!injected) {
				throw new Error('Injection failed');
			}

			result = injected;
		}

		//#region Check generated and/or injected proptypes
		const propTypes = prettier.format(result, {
			...prettierConfig,
			filepath: outputPath,
		});

		if (fs.existsSync(outputPath)) {
			expect(propTypes.replace(/\r?\n/, '\n')).toMatch(
				fs.readFileSync(outputPath, 'utf8').replace(/\r?\n/, '\n')
			);
		} else {
			fs.writeFileSync(outputPath, propTypes);
		}
		//#endregion
	});
}
