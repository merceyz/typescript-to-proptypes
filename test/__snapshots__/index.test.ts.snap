// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs 1`] = `
Object {
  "body": Array [
    Object {
      "name": "Foo",
      "type": "ComponentNode",
      "types": Array [
        Object {
          "jsDoc": undefined,
          "name": "bar",
          "propType": Object {
            "type": "StringNode",
          },
          "type": "PropTypeNode",
        },
      ],
    },
  ],
  "type": "ProgramNode",
}
`;
